policy_agent/policy_agent.py

```diff
@@ -0,0 +1,354 @@
+            r_kind = r_obj["kind"]
+
+            _log.info("%s: %s" % (r_type, r_kind))
+
+            if r_kind in ["Pod", "Namespace", "Service"]:
+                self.process_resource(action=r_type, kind=r_kind, target=r_obj)
+            else:
+                _log.info(
+                    "Resource %s not Pod, Service, or Namespace" % r_kind)
+
+        except Queue.Empty:
+            self.resync()
+
+    def process_resource(self, action, kind, target):
+        """
+        Takes a target object and an action and updates internal resource pools
+        :param action: String of ["ADDED", "DELETED", "MODIFIED"] (returned by api watch)
+        :param kind: String of ["Pod", "Namespace", "Service", "Endpoints"]
+        :param target: json dict of resource info
+        """
+        # Determine Resource Pool 
+        if kind == "Pod":
+            resource_pool = self.pods
+            obj = Pod(target)
+
+        elif kind == "Service":
+            resource_pool = self.services
+            obj = Service(target)
+
+        elif kind == "Endpoints":
+            resource_pool = self.endpoints
+            obj = Endopints(target)
+
+        elif kind == "Namespace":
+            resource_pool = self.namespaces
+            obj = Namespace(target)
+
+        else:
+            _log.error("resource %s not Pod, Service, or Namespace" % kind)
+            sys.exit(1)
+
+        target_key = obj.key
+
+        if action == "ADDED":
+            if target_key not in resource_pool:
+                resource_pool[target_key] = obj
+                _log.info("%s added to Calico store" % target_key)
+
+            else:
+                _log.error("Tried to Add %s, but %s already in bin" %
+                           (obj, target_key))
+
+        elif action == "DELETED":
+            if target_key in resource_pool:
+                del resource_pool[target_key]
+                _log.info("%s deleted from Calico store" % target_key)
+
+            else:
+                _log.error(
+                    "Tried to Delete %s, but %s was not in bin" %
+                    (obj, target_key))
+
+        elif action == "MODIFIED":
+            if target_key in resource_pool:
+                _log.info("Updating %s\n%s\n=====>\n%s" %
+                          (target_key, resource_pool[target_key], obj))
+                resource_pool[target_key] = obj
+
+            else:
+                _log.warning("Tried to Modify %s, but %s was not in bin. "
+                             "Treating as Addition" %
+                             (target_key, target_key))
+                self.process_resource(action="ADDED", kind=kind, target=target)
+
+    def resync(self):
+        """
+        Tells all resource objects to resync their profile information
+        """
+        for resource_pool in [self.pods, self.services, self.endpoints, self.namespaces]:
+            for resource_key in resource_pool:
+                resource_pool[resource_key].resync()
```

Comment:
I don't think pushing the resync logic into the individual "resource" objects is the right move here.  It is going to make it difficult to combine information from multiple types of objects, and makes the resync code much harder to read and debug.

Let's talk about what the right abstraction is here.